src/modules/battery.rs

```diff
@@ -14,60 +16,60 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
     let BatteryStatus { state, percentage } = battery_status;
 
     let mut module = context.new_module("battery");
-    let battery_config: BatteryConfig = BatteryConfig::try_load(module.config);
+    let config: BatteryConfig = BatteryConfig::try_load(module.config);
 
-    // Parse config under `display`
-    let display_styles = &battery_config.display;
-    let display_style = display_styles
+    // Parse config under `display`.
+    // Select the first style that match the threshold,
+    // if all thresholds are lower do not display battery module.
+    let display_style = config
+        .display
         .iter()
-        .find(|display_style| percentage <= display_style.threshold as f32);
+        .find(|display_style| percentage <= display_style.threshold as f32)?;
 
-    if let Some(display_style) = display_style {
-        // Set style based on percentage
-        module.set_style(display_style.style);
-        module.get_prefix().set_value("");
+    // Parse the format string and build the module
+    match StringFormatter::new(config.format) {
+        Ok(formatter) => {
+            let formatter = formatter
+                .map_style(|style| match style {
+                    "style" => Some(Ok(display_style.style)),
+                    _ => None,
+                })
+                .map(|variable| match variable {
+                    "percentage" => Some(Ok(format!("{}{}", percentage.round(), percentage_char))),
+                    _ => None,
+                })
+                .map(|variable| match variable {
+                    "symbol" => match state {
+                        battery::State::Full => Some(Ok(config.full_symbol)),
+                        battery::State::Charging => Some(Ok(config.charging_symbol)),
+                        battery::State::Discharging => Some(Ok(config.discharging_symbol)),
+                        battery::State::Unknown => Some(Ok(config.unknown_symbol)),
+                        battery::State::Empty => Some(Ok(config.empty_symbol)),
+                        _ => {
+                            log::debug!("Unhandled battery state `{}`", state);
+                            None
+                        }
+                    },
```

Comment:
I think `symbol` should be configured as a format string instead of a plain string.

It should be fine most of the case but there may be some users trying to configure a rainbow colored `FULL` or so.

Note that `map_meta` accepts a format string as a return value.